### ISSUE TITLE
`tox-sink` custom location within DOM

### ISSUE THREAD

**Issue by mspiderv** [tinymce/tinymce#7868]
It would be great if I could customize where the `<div class="tox tox-silver-sink tox-tinymce-aux" style="position: relative;"></div>` will be rendered. By default it's rendered at the bottom of the `<body>` but I would like to change this location.

Is this possible?

**Comment by TheSpyder** [comment]
The problem with moving it is that can impact the visibility of dialogs and menus - these elements generally need to extend outside of the editor bounds. By forcing the sink to be at the end of the body it's easier to make sure those are on top of all other content.

We are planning to look at how to support scenarios where this causes issues, such as when TinyMCE is in a scrollable container. Can you provide more details about why you'd like to move the sink?

**Comment by mspiderv** [comment]
Im building an application based on [Quasar](https://quasar.dev/) (Vue 3). Quasar implements it's own dialog called `QDialog`. This dialog works in such a way, that it keeps focus inside the dialog (you cannot focus any element outside the dialog). So for example I have an `<input>` that is focused somewhere in my application. When I open a new QDialog, Quasar automatically focuses the first focusable element inside the dialog. When I try to focus the original `<input>` that is not inside the dialog (using mouse or keyboard) Quasar detects this and automatically refocus the first focusable element inside the QDialog. This all is expected behaviour, there is no problem with that, because user should not be able to work with other parts of the UI when some dialog is opened.

The problem is that Quasar keeps focus **INSIDE** the dialog. So when I put my Tinymce inside Quasar's dialog, then Im not able to focus anythig that is located inside Tinymce's dialog. The technical reason for this is that Tinymce's dialogs are located under `<body>` so these dialogs **ARE NOT** a child elements of Quasar's dialog. So Quasar consider Tinymce's dialogs as somethig that is not a part of Tinymce editor and therefore Im not able to interact with Tinymce's dialogs.

[Here is a working example](https://codesandbox.io/s/sharp-architecture-8bfu6p?file=/src/pages/Index.vue)

You can click on "SHOW DIALOG". A new QDialog with Tinymce inside will open.

Then try to add a link.
![image](https://user-images.githubusercontent.com/2259835/171832356-e5d5b9e7-62ac-49e8-87ea-a5f222e93660.png)

Try to focus an `<input>` inside the TinyMCE's dialog. Quasar will not allow you to do that, because as I said, this dialog is located under `<body>` so Quasar will always refocus the first focusable element withtin the Quasar's dialog.
![image](https://user-images.githubusercontent.com/2259835/171832513-812598a1-8db2-4504-9351-70770db7081e.png)

If there would be an option to customize where the `div.tox-sink` is located, I would put it inside the QDialog and fix `z-index` issues on my own.

**Comment by TheSpyder** [comment]
Thank you for the detailed update! I understand what you mean. I will add your example to our notes.

I tried manually moving the sink in your example (using dev tools) and z-index wasn't really an issue, Quasar's capturing of mouse events was. But it did at least prove the concept might work so long as we note in the documentation that these sorts of issues will need to be managed by the developer.

**Comment by tobiashm** [comment]
Another example is if you use TinyMCE inside the native `<dialog>` element, then anything rendered in the sink will always be shown behind the dialog, since it's located in the ["top layer"](https://developer.chrome.com/blog/what-is-the-top-layer/) so no amount of z-index will help.

In this case it would be nice if we could specify that the sink should be places inside the `<dialog>` element instead of the `<body>`.

**Comment by bqluan** [comment]
Quasar's <q-dialog> has a prop `allow-focus-outside` add at v2.7.2.

**Comment by mspiderv** [comment]
@bqluan Yes. Guess who asked Quasar developers to add that feature :) https://github.com/quasarframework/quasar/issues/13595

**Comment by mpskovvang** [comment]
I'm experiencing an issue, that also seems impossible to solve with only `z-index`.

In my case, we have an editor in the main content area and an editor shown in a modal. Setting aux `z-index` higher than the modal ensures menus etc. are shown on top. However, notifications in the main editor share the same `z-index` and are therefore also shown on top of the modal.

![image](https://user-images.githubusercontent.com/7034075/219555609-8d73c298-14a4-4e07-9927-020576d1a579.png)

Placing the aux element closer to the editor (in this case in the modal), would support `z-index` stacking, but I see how this could cause issues related to positioning/overflow.

Any suggestions?

**Comment by TheSpyder** [comment]
We are working on a solution that moves the sink next to the editor - #5097 

I'm not sure if that will fix this problem, so I'll leave it open for now, but we aren't likely to support completely custom sink locations.

**Comment by mpskovvang** [comment]
Thanks, I'll see if it makes any changes when released.

In the meantime, I made a workaround with a custom static UI element.

![image](https://user-images.githubusercontent.com/7034075/219590870-ec739ac1-1d5f-48eb-b9c9-e5226aaf8b6a.png)

**Comment by ericbf** [comment]
Using `ui_mode: "split"` will put the sink right after the editor, which should put it inside your dialog.

**Comment by TheSpyder** [comment]
yes, that was released in TinyMCE 6.4 and I forgot to follow up.

**Comment by kyh196201** [comment]
> Using `ui_mode: "split"` will put the sink right after the editor, which should put it inside your dialog.

@ericbf it works for me thank you !!! 👏👏

**Comment by MagnusOxlund** [comment]
> I tried manually moving the sink in your example (using dev tools) and z-index wasn't really an issue, Quasar's capturing of mouse events was. But it did at least prove the concept might work so long as we note in the documentation that these sorts of issues will need to be managed by the developer.

I think it would be great if you could mention the "sink" in the documentation. I'm new to TinyMCE, and it took quite a bit of digging through the source and other issues to find out what the sink is and what it's responsible before before I could fix the layout issues it caused (vanilla CSS grid-based) without affecting the editor.

**Comment by rodhoward** [comment]
I'm using the `ui_mode: "split" ` and its working for most things tables and dropdowns. We are using native dialogs where everything in the dialog is rendered on the top level. 

The problem is still happening for plugins that open their own dialog.  For example the link:
![image_720](https://github.com/user-attachments/assets/bd2e9018-fd3c-4bf7-bd65-ed737e74ad6c)
Here you see the dialog element is still in the main body not inside the parent dialog which is causing it to render behind and is not actually clickable.

**Comment by TheSpyder** [comment]
Yes, `ui_mode: "split"` only applies to elements that need to scroll with the editor. Modal dialogs that intend to take over the screen are still put into the global sink to help ensure they can break out of restrictions applied to the editor container.

You could try increasing the `z-index` of `.tox-tinymce-aux` to bring the TinyMCE dialogs in front of your dialog that contains TinyMCE.

Long term, rather than continue to patch this old system there are modern features like [dialog](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog) and [popover](https://developer.mozilla.org/en-US/docs/Web/API/Popover_API) we could look at to display our UI. However we don't have any firm plans for this.

**Comment by lbird-mma** [comment]
Guess I'll just add some context here too. 

I am using Shoelace as a framework for my webproject. For help editing purposes I've set up TinyMCE within a Shoelace drawer, which is a modal display. I couldn't get Tiny's link editor or source editor to work, as this happens in a new modal. The modal of Shoelace traps any focus.

I submitted my problem to Shoelace's help. They have a solution for this in place, which is `modal.activateExternal()` to temporarily disable focus trapping. Someone else already ran into the [same problem there](https://github.com/shoelace-style/shoelace/issues/1571).

To demonstrate I worked out a non-working and working situation in [Codepen](https://codepen.io/lbird-mma/pen/raBXxQb).

For now I'm saved, but the Shoelace project is moving forward to Web Awesome where, instead of a faux modal, they will be using a `<dialog>`. For this to work in conjunction it would be necessary that TinyMCE starts using the Popover API. The `<dialog>` elements are becoming quite ubiquitous anyway.

So any plans to move in that direction I would applaud to !

**Comment by bes-internal** [comment]
We are gradually moving to `<dialog>` and have problems with TinyMCE because its dialogs are not in ::top-layer.
`ui_mode: "split"` doesn't help because it doesn't work in all cases, and in ours it never does
rel. #5525

**Comment by bes-internal** [comment]
see https://github.com/tinymce/tinymce/issues/5525#issuecomment-1497091996
